cli/src/main/java/org/aya/cli/library/incremental/InMemoryCompilerAdvisor.java

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2020-2022 Yinsen (Tesla) Zhang.
+// Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
+package org.aya.cli.library.incremental;
+
+import kala.collection.immutable.ImmutableSeq;
+import kala.collection.mutable.MutableMap;
+import org.aya.cli.library.source.LibraryOwner;
+import org.aya.cli.library.source.LibrarySource;
+import org.aya.core.def.GenericDef;
+import org.aya.core.serde.SerTerm;
+import org.aya.core.serde.Serializer;
+import org.aya.resolve.ResolveInfo;
+import org.aya.resolve.module.ModuleLoader;
+import org.aya.util.reporter.Reporter;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileTime;
+
+public record InMemoryCompilerAdvisor(
+  @NotNull MutableMap<Path, FileTime> coreTimestamp,
+  @NotNull MutableMap<ImmutableSeq<String>, ResolveInfo> compiledCore
+) implements CompilerAdvisor {
+  public InMemoryCompilerAdvisor() {
+    this(MutableMap.create(), MutableMap.create());
+  }
+
+  @Override public boolean isSourceModified(@NotNull LibrarySource source) {
+    var coreLastModified = coreTimestamp.getOption(source.file());
+    try {
+      if (coreLastModified.isEmpty()) return true;
+      return Files.getLastModifiedTime(source.file())
+        .compareTo(coreLastModified.get()) > 0;
+    } catch (IOException ignore) {
+      return true;
+    }
+  }
+
+  @Override public void updateLastModified(@NotNull LibrarySource source) {
+    try {
+      coreTimestamp.put(source.file(), Files.getLastModifiedTime(source.file()));
+    } catch (IOException ignore) {
```

Comment:
Are you sure you wanna ignore that?

Reply:
It's safe to ignore. If the `getLastModifiedTime` failed, the `coreTimestamp` is not updated (the core file still has a smaller modified time than source file) so the corresponding source will be re-compiled next time even if it is really up-to-date.

Reply:
How about only putting getLastModifiedTime in the try block?

Reply:
> How about only putting getLastModifiedTime in the try block?

In this case if `getLastModifiedTime` failed, we still have nothing to put. So I guess this does not matter.